hail/python/hailtop/aiotools/fs.py

```diff
@@ -0,0 +1,164 @@
+from concurrent.futures import ThreadPoolExecutor
+import urllib.parse
+from hailtop.utils import blocking_to_async
+from .stream import AsyncStream, blocking_stream_to_async
+
+AsyncFSType = TypeVar('AsyncFSType', bound='AsyncFS')
+
+
+class AsyncFS(abc.ABC):
+    @abc.abstractmethod
+    def schemes(self) -> Set[str]:
+        pass
+
+    @abc.abstractmethod
+    async def open(self, url: str, mode: str = 'r') -> AsyncStream:
+        pass
+
+    @abc.abstractmethod
+    async def mkdir(self, url: str) -> None:
+        pass
+
+    @abc.abstractmethod
+    async def isfile(self, url: str) -> bool:
+        pass
+
+    @abc.abstractmethod
+    async def isdir(self, url: str) -> bool:
+        pass
+
+    @abc.abstractmethod
+    async def remove(self, url: str) -> None:
+        pass
+
+    @abc.abstractmethod
+    async def rmtree(self, url: str) -> None:
+        pass
+
+    async def touch(self, url: str) -> None:
+        async with await self.open(url, 'wb'):
+            pass
+
+    async def close(self) -> None:
+        pass
+
+    async def __aenter__(self: AsyncFSType) -> AsyncFSType:
+        return self
+
+    async def __aexit__(self,
+                        exc_type: Optional[Type[BaseException]],
+                        exc_val: Optional[BaseException],
+                        exc_tb: Optional[TracebackType]) -> None:
+        await self.close()
+
+
+class LocalAsyncFS(AsyncFS):
+    def __init__(self, thread_pool: ThreadPoolExecutor, max_workers=None):
+        if not thread_pool:
+            thread_pool = ThreadPoolExecutor(max_workers=max_workers)
+        self._thread_pool = thread_pool
+
+    def schemes(self) -> Set[str]:
+        return {'file'}
+
+    @staticmethod
+    def _get_path(url):
+        parsed = urllib.parse.urlparse(url)
+        if parsed.scheme and parsed.scheme != 'file':
+            raise ValueError(f"invalid scheme, expected file: {parsed.scheme}")
+        return parsed.path
+
+    async def open(self, url: str, mode: str = 'r') -> AsyncStream:
+        if 'b' not in mode:
+            raise ValueError(f"can't open: text mode not supported: {mode}")
+        return blocking_stream_to_async(self._thread_pool, cast(BinaryIO, open(self._get_path(url), mode)))
+
+    async def mkdir(self, url: str) -> None:
+        path = self._get_path(url)
+        os.mkdir(path)
+
+    async def isfile(self, url: str) -> bool:
+        path = self._get_path(url)
+        return await blocking_to_async(self._thread_pool, os.path.isfile, path)
+
+    async def isdir(self, url: str) -> bool:
+        path = self._get_path(url)
+        return await blocking_to_async(self._thread_pool, os.path.isdir, path)
+
+    async def remove(self, url: str) -> None:
+        path = self._get_path(url)
+        return os.remove(path)
+
+    async def rmtree(self, url: str) -> None:
+        path = self._get_path(url)
+        await blocking_to_async(self._thread_pool, shutil.rmtree, path)
+
+
+class RouterAsyncFS(AsyncFS):
```

Comment:
Why is this called a RouterAsyncFS and takes a list of filesystems? This seems like a wrapper for all file systems and should just have all possible file systems as a class property and not take the list of filesystems as an argument.

Reply:
We will not always be able to construct all filesystems in all deployments.  For example, the Google Storage FS fails if it cannot find credentials.

Reply:
Also, I'm not a fan of totally global configuration.  The Hadoop design for example caused us lots of problem and should be avoided.